### PR TITLE
[feature/app-provider] App Provider support

### LINKED ISSUE #1127: [FEATURE REQUEST] "Open in Web" for native clients
**Story**
As a client user I want to open a file in the Web Office directly from the context menu so that I can easily start collaborating with others

**Specifics**
- Infinite Scale only
- Use App provider backend
- Timeline ~ mid August

@hosy @felix-schwarz

### PR DESCRIPTION
**Pull request by felix-schwarz**
## Description
Adds support for opening files in web apps and for creating new files using the App Provider API.

## Related Issue
#1127

## Developer Documentation
- https://owncloud.dev/services/app-registry/apps/

## Testing
The three modes (default browser, in-app, in-app with default browser option) can be set via MDM, `Branding.plist` and env variables.

The Xcode project's scheme already has ready-to-use that can be activated prior to launch: 
Bildschirmfoto 2022-09-26 um 16 27 09

Test instance with WOPI apps installed: http://ocis.ocis-wopi.latest.owncloud.works

### Creating new documents
Empty folder action | Folder action | File type picker | File name entry | New file
--|--|--|--|--
Simulator Screen Shot - iPhone 13 mini - 2022-09-08 at 11 47 57 | Simulator Screen Shot - iPhone 13 mini - 2022-09-08 at 11 46 38 | Simulator Screen Shot - iPhone 13 mini - 2022-09-08 at 11 48 02 | Simulator Screen Shot - iPhone 13 mini - 2022-09-08 at 11 48 05 | Simulator Screen Shot - iPhone 13 mini - 2022-09-08 at 11 48 09

### Open document in web app
In info sheet | In context menu | Opened in Safari | In-app | In-app, with default browser option
---|---|---|--|--
Simulator Screen Shot - iPhone 13 mini - 2022-09-08 at 11 48 16 | Simulator Screen Shot - iPhone 13 mini - 2022-09-08 at 11 52 21 | Simulator Screen Shot - iPhone 13 mini - 2022-09-08 at 11 48 27 | Simulator Screen Shot - iPhone 13 mini - 2022-09-19 at 12 23 30 | Simulator Screen Shot - iPhone 13 mini - 2022-09-19 at 12 22 39

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## QA

Test plan:

- [x] (1) Close button not visible in dark themes https://github.com/owncloud/ios-app/pull/1151#issuecomment-1265348464
- [x] (2) Feedback from app with lack of connection https://github.com/owncloud/ios-app/pull/1151#issuecomment-1265362403
- [x] (3) Lack of feedback in new document creation (offline) https://github.com/owncloud/ios-app/pull/1151#issuecomment-1265381295
- [x] (4) Default browser only the first time https://github.com/owncloud/ios-app/pull/1151#issuecomment-1266536491

### CODE REVIEW

**Comment by jesmrec** [comment]
### (1)

Close button is almost invisible in Dark themes, with inline editor:

Screenshot 2022-10-03 at 14 08 45

**Comment by jesmrec** [comment]
### (2)

1. Download an OpenOffice file
2. Remove device connection
3. Select the `Open in OnlyOffice (web)` over the downloaded file

Current: nothing happens. 
Expected: some error feedback

iPhoneXR 16.0.2
`4b9af63aa`

**Comment by jesmrec** [comment]
### (3) 

1. Remove device connection
2. Select `+` and `New document`
3. Enter a correct name
4. Submit

Current: Nothing happens (should be file created and waiting for connection using the offline mode?). After recovering connection, status bar shows `95% completed` forever:

https://user-images.githubusercontent.com/14894746/193578292-a0d99225-10eb-46e0-89e8-61c30e73d748.MP4

iPhoneXR 16.0.2
iPadAir 15.7

`4b9af63aa`

**Comment by jesmrec** [comment]
### (4)

1. Before building the app, select the option `default-browser`, to open the files in the external one (Safari)

Screenshot 2022-10-04 at 09 39 35

2. Select `Open in xxx` -> file opened in Safari browser correctly
3. Kill oC app and Safari app
4. Reopen oC app
5. Repeat step 2.

Current: file opened in the inline browser
Expected: file opened in Safari (or default browser)

iPhoneXR 16.0.2
iPadAir 15.7

`4b9af63aa`

**Comment by hosy** [comment]
## (5) Missing Translation

Some translation strings are missing:
https://github.com/owncloud/KoKo22/issues/85

**Comment by felix-schwarz** [comment]
@jesmrec (1), (2) and (3) are fixed as of https://github.com/owncloud/ios-app/pull/1151/commits/e8b125ec3dd0efbba0540105fa2a3c1ae2e20b2c

(4) is not a bug. The selected options in Xcode control which env variables Xcode sets for the app when it launches it. They don't persist and will only apply to the launched process. So when you relaunch the app outside Xcode, those env vars are gone and no longer apply.

@hosy (5) I looked through the strings and added those that previously weren't in `Localizable.strings`.

**Comment by jesmrec** [comment]
(3) fixed in https://github.com/owncloud/ios-app/pull/1153

**Comment by jesmrec** [comment]
(3) was fixed in [12.0 experimental branch](https://github.com/owncloud/ios-app/pull/1153), so, if the current PR is merged somewhere, the fix will be lost unless experimental is merged too. Please, take this is account. 

Feature is totally approved in experimental branch.

**Comment by felix-schwarz** [comment]
@je

> (3) was fixed in [12.0 experimental branch](https://github.com/owncloud/ios-app/pull/1153), so, if the current PR is merged somewhere, the fix will be lost unless experimental is merged too. Please, take this is account.
> 
> Feature is totally approved in experimental branch.

That fix is in `feature/new-navigation`.

**Comment by jesmrec** [comment]
> That fix is in feature/new-navigation.

Ok, just be careful if the current branch is merged somewhere, in order not to lose that fix. I let it in your hands.

Approved.

**Review by jesmrec — APPROVED** [review]

File: ownCloud/Client/Actions/Actions+Extensions/CreateDocumentAction.swift [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+
+					OnMainThread {
+						let documentNameViewController = NamingViewController( with: self.core, defaultName: suggestedName, stringValidator: { name in
+							if name.contains("/") || name.contains("\\") {
+								return (false, nil, "File name cannot contain / or \\".localized)
+							} else {
+								if let item = item {
+									if ((try? self.core?.cachedItem(inParent: item, withName: name, isDirectory: true)) != nil) ||
+									   ((try? self.core?.cachedItem(inParent: item, withName: name, isDirectory: false)) != nil) {
+										return (false, "Item with same name already exists".localized, "An item with the same name already exists in this location.".localized)
```
Comment (hosy) [inline ownCloud/Client/Actions/Actions+Extensions/CreateDocumentAction.swift:117]: Localisation is missing in strings file

File: ownCloud/Client/Actions/Actions+Extensions/OpenInWebAppAction.swift [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+					]
+				]
+			])
+		}
+
+		return super.actionExtension
+	}
+
+	override class var identifier : OCExtensionIdentifier? { return OCExtensionIdentifier("com.owncloud.action.openinwebapp") }
+	override class var category : ActionCategory? { return .normal }
```
Comment (hosy) [inline ownCloud/Client/Actions/Actions+Extensions/OpenInWebAppAction.swift:70]: keyboard command is missing

**Review by hosy — CHANGES_REQUESTED** [review]

File: ownCloud/Client/Actions/Actions+Extensions/CreateDocumentAction.swift (hunk 1 above)
Comment (felix-schwarz) [reply ownCloud/Client/Actions/Actions+Extensions/CreateDocumentAction.swift:117]: This was added in `feature/new-navigation`.

**Comment by felix-schwarz** [comment]
Merged into `milestone/12.0` in 0a80962b27de768c31523233c75e8693d84d7dd3.

**Pull request closed without merging**